doc/high_availability.md

```diff
@@ -1,110 +1,74 @@
- * The transaction resumer
- * The minunits worker
-
-When a machine agent connects to the state, it decides whether
-it is on the same instance as the mongo master instance, and
-if so, it runs the singleton workers; otherwise it doesn't run them.
-
-Because we are using `mgo.Strong` consistency semantics,
-it's guaranteed that our mongo connection will be dropped
-when the master changes, which means that when the
-master changes, the machine agent will reconnect to the
-state and choose whether to run the singleton workers again.
-
-It also means that we can never accidentally have two
-singleton workers performing operations at the same time.
+HA mode is invoked via the `enable-ha` command. By default, this will ensure 
+three controllers, but using the `-n` flag will allow running with five or 
+seven controllers.
+
+The number of controllers can be reduced by invoking `remove-machine` with a 
+controller machine ID, and increased by re-running the `enable-ha` command.
+
+### Dqlite
+
+Each controller is a Dqlite node. The `dbaccessor` worker on each controller is 
+responsible for maintaining the Dqlite cluster. When entering HA mode, the 
+`dbaccessor` worker will configure the local Dqlite node as a member of the 
+cluster.
+
+When starting Dqlite, the worker must bind it to an IP address on the local
+machine. It is a requirement that there is a unique local-cloud-scoped address
+for Dqlite to use. If there is no unique address, new nodes will not be joined 
+to the cluster until one can be determined. See _Controller Charm_ below.
+
+Beyond joining nodes to the cluster, Juju does not manage Dqlite node roles.
+This is handled within Dqlite itself. A cluster will have one leader, voters
+which are eligible to participate in leader elections, stand-bys and spares.
+
+Juju does not predicate any logic on node roles.
+
+If the number of controller instances is reduced to one, the `dbaccessor` 
+worker will recognise this scenario and proactively reconfigure the cluster to
+be constituted by the local node only.
+
+### Controller Charm
+
+The controller charm propagates binding information to the `dbaccessor` worker.
+It coordinates with other controller units via the `db-cluster` peer relation.
+If there are multiple potential bind addresses for a Dqlite node, the user must
+supply an endpoint binding for the relation using a space that ensures a unique
+IP address.
+
+### API Addresses for Agents
+
+When machines in the control plane change, agent configuration files are 
+re-written with usable API addresses from all controllers. Agents will try
+these addresses in random order when connecting, so that an inaccessible 
+controller just results in connection to another.
+
+The list of addresses supplied to agent configuration can be influenced by the
+`juju-ctrl-space` controller configuration article. This is supplied with a
+space name in order that agent-controller communication can be isolated to 
+specific networks.
```

Comment:
I'm pretty sure this is still spelled `juju-mgmt-space` in 4.0. Ideally it would also become a binding rather than a controller configuration. But since we don't model the relations of each agent to the controller, it doesn't really have a place to live. It *is* what `extra-bindings` was designed for, but I know we don't like the nomenclature. With the caveat that all progress on *improving* the network model has stalled, which means we at least have *something* that can be used, even if it isn't perfect yet.

https://github.com/juju/juju/blob/d0d39e6e8ee41039b1befa728c03db19b6cbd4ab/controller/config.go#L240